Error reporting

sallybg opened the pull request:
Make pre-map and post-map objects optional in final JSON output. Report per-variant and per-score-set errors. Add versioning.

sallybg committed:
Initialize dcd_mapping api server

Co-authored-by: Ben Capodanno <31941502+bencap@users.noreply.github.com>

sallybg committed:
Have the api create vrs mapped objects with preferred layer only

Co-authored-by: Ben Capodanno <31941502+bencap@users.noreply.github.com>

sallybg committed:
squash: formatting changes

sallybg committed:
Allow pre-map genomic variant without corresponding post-map genomic variant

Currently, the only case where this will occur is when the variant falls
outside of the region of the target sequence that aligns to the genomic
reference sequence. Post-map protein variants are still required to
output the corresponding pre-map protein variant, because we don't
expect a protein variant to fall outside the region of the target
sequence that aligns to the reference transcript, although
this assumption should be evaluated as a separate issue.

sallybg committed:
Generate MappedScore with error message if mapping fails

sallybg committed:
fixup mapped error messages

sallybg committed:
Add score set-based error handling

sallybg committed:
Initialize versioning, add date and version to output

sallybg committed:
Add error handling for external data sources

sallybg committed:
Add error messages for VRS id lookups

sallybg committed:
Merge branch 'api' into error-reporting

sallybg committed:
Add error handling to api map function

sallybg committed:
Add API docstrings

bencap commented on src/dcd_mapping/__init__.py:
```diff
@@ -8,7 +8,9 @@
+from .version import dcd_mapping_version
 
 __all__ = ["map_scoreset", "map_scoreset_urn"]
+__version__ = dcd_mapping_version
```
Is there a reason we can't just define the version here rather than maintaining a separate file only for the version?

bencap commented on src/dcd_mapping/align.py:
```diff
@@ -205,6 +205,7 @@ def _get_best_hit(output: QueryResult, urn: str, chromosome: str | None) -> Hit:
         else:
             if list(output):
                 hit_chrs = [h.id for h in output]
+                # TODO should this be an error rather than a warning? it seems like a problem if we can't find a hit on the expected chromosome
```
My sense is that this is fine as a warning since the chromosome isn't explicitly provided by the user, and is instead inferred via score set metadata.

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -251,34 +251,35 @@ def _annotate_allele_mapping(
+            sequence_id = f"ga4gh:{mapped_score.post_mapped.location.sequenceReference.refgetAccession}"
+            accession = get_chromosome_identifier_from_vrs_id(sequence_id)
+            if accession is None:
+                raise ValueError
```
Could we add a message to these raised ValueErrors? Even if we don't expect them to occur it's easier to read what went wrong while looking at the flow.

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -251,34 +251,35 @@ def _annotate_allele_mapping(
+                accession = accession[7:]
+        else:
+            if tx_results is None:
+                raise ValueError  # impossible by definition
```
> Could we add a message to these raised ValueErrors? Even if we don't expect them to occur it's easier to read what went wrong while looking at the flow.

Same as above

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -318,28 +322,32 @@ def _annotate_haplotype_mapping(
+        mavedb_id=mapped_score.accession_id,
+        score=float(mapped_score.score) if mapped_score.score is not None else None,
+        annotation_layer=mapped_score.annotation_layer,
+        error_message=mapped_score.error_message
+        if mapped_score.error_message
+        else None,  # TODO might not need if statement here
```
> Based on the way the model is defined, I agree you shouldn't need the if statement.

Same as above.

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -287,24 +288,27 @@ def _annotate_allele_mapping(
         mavedb_id=mapped_score.accession_id,
         score=float(mapped_score.score) if mapped_score.score else None,
         annotation_layer=mapped_score.annotation_layer,
+        error_message=mapped_score.error_message
+        if mapped_score.error_message
+        else None,  # TODO might not need if statement here
```
Based on the way the model is defined, I agree you shouldn't need the if statement.

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -367,23 +375,34 @@ def annotate(
                 )
             )
         else:
+            # TODO how to combine this error message with other potential error messages?
             ValueError("inconsistent variant structure")
```
Not sure if these will be other exceptions, or just messages, but we could try doing something with exception groups if this was still an issue: https://docs.python.org/3/tutorial/errors.html#raising-and-handling-multiple-unrelated-exceptions

Also, is this missing a `raise` or is that intentional?

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -507,10 +553,22 @@ def save_mapped_output_json(
         reference_sequences[layer][
             "mapped_reference_sequence"
         ] = _get_mapped_reference_sequence(layer, tx_output, align_result)
+    # except Exception as e:
+    #     _logger.warning(
+    #         str(e)
+    #     )
+    #     output = ScoresetMapping(
+    #         metadata=metadata,
+    #         error_message = str(e).strip("'")
+    #     )
+
+    #     return write_scoreset_mapping_to_json
```
```suggestion
```
Can be removed

bencap commented on src/dcd_mapping/lookup.py:
```diff
@@ -212,12 +212,15 @@ async def get_protein_accession(transcript: str) -> str | None:
+        """  # noqa: S608
+        result = await uta.execute_query(query)
+    except Exception as e:
+        raise DataLookupError from e
```
Should we surface a message here?

bencap commented on src/dcd_mapping/lookup.py:
```diff
@@ -239,16 +242,19 @@ async def get_transcripts(
+        """  # noqa: S608
+        result = await uta.execute_query(query)
+    except Exception as e:
+        raise DataLookupError from e
```
> Should we surface a message here?

Same as above

bencap commented on src/dcd_mapping/vrs_map.py:
```diff
@@ -633,12 +675,13 @@ def vrs_map(
-            click.echo(msg)
-        _logger.warning(msg)
-        return None
+    # TODO address this hardcoding, and if we keep it, this should be a score set mapping error message
```
Is this appropriately handled now? If so, we can remove the commented block.

bencap commented on src/dcd_mapping/schemas.py:
```diff
@@ -164,12 +167,15 @@ class ScoreAnnotation(BaseModel):
     mavedb_id: StrictStr
-    relation: Literal["SO:is_homologous_to"] = "SO:is_homologous_to"
-    score: float | None
+    relation: Literal[
+        "SO:is_homologous_to"
+    ] = "SO:is_homologous_to"  # TODO this should probably be None if pre_mapped is false?
```
That makes sense to me. Does pre-mapped being false imply that people should just ignore this? Is it a lot of effort to change?

bencap commented on src/dcd_mapping/vrs_map.py:
```diff
@@ -374,40 +408,78 @@ def _map_genomic(
+    if (
+        row.hgvs_nt in {"_wt", "_sy", "="}
+        or "fs"
+        in row.hgvs_nt  # TODO I think this line can be taken out, I don't think fs nomenclature can be used for nt anyway
```
Were there any score sets you encountered where this was a problem? I am not well positioned to answer, but figured I'd comment in case we should discuss with someone.

bencap reviewed (commented):
Thanks for all the work on this Sally, going to be great to get this up soon. Most of these are pretty minor and focused around the added TODOs or surfaced errors. Flow/logic/design all seemed good to me.

sallybg committed:
Use Field() to set values for default fields in ScoresetMapping

Since ScoresetMapping is returned as an API response, the typical
way of setting default values does not work.

bencap committed:
Make Authenticated Requests if Envvars are Set

sallybg committed:
User python serialization mode instead of json

FastAPI uses json serialization by default, which flattens the Allele
sequence location field. To avoid this, use pydantic's serializer with python mode
instead, and return the response directly as a JSON response.

bencap committed:
Set envvar in Git Actions Tests
